src/adios2_io.cpp

```diff
@@ -0,0 +1,163 @@
+// ex. ./bpls mesh.bp
+
+int main(int argc, char *argv[])
+{
+  auto lib = Omega_h::Library(&argc, &argv);
+  auto world = lib.world();
+
+  if (lib.world()->size()>1)
+  {
+    if (!lib.world()->rank())
+      fprintf(stderr, "ADIOS2 file I/O with partitioned mesh is not supported yet\n");
+    exit(EXIT_FAILURE);
+  }
+
+  Omega_h::CmdLine cmdline;
+
+  cmdline.add_arg<std::string>("input.osh");
+  cmdline.add_arg<std::string>("output.bp");
+  if (!cmdline.parse_final(world, &argc, argv)) return -1;
+  auto inpath = cmdline.get<std::string>("input.osh");
+  auto outpath=cmdline.get<std::string>("output.bp");
+
+  Omega_h::Mesh mesh(&lib);
+  Omega_h::binary::read(inpath, world, &mesh);
+  cout<<"\n--- Mesh loaded from \""<<inpath<<"\" ---\n";
+  print_info(&lib, mesh);
+
+//  Omega_h::Mesh mesh = build_box(world, OMEGA_H_SIMPLEX, 1., 1., 0., 2, 2, 0);
+  Omega_h::binary::write("omegah.osh", &mesh);
+  Omega_h::vtk::write_parallel("omegah.vtk", &mesh);
+
+  try
+  {
+    write_adios2(outpath, &mesh);
+    Omega_h::Mesh mesh2 = read_adios2(outpath, &lib);
+    Omega_h::vtk::write_parallel("adios2.vtk", &mesh2);
+
+    cout<<"\n\n--- Mesh loaded back from \""<<outpath<<"\" ---\n";
+    print_info(&lib, mesh2);
+
+    double tol = 1e-6, floor = 0.0;
+    bool allow_superset = false;
+    auto opts = MeshCompareOpts::init(
+                &mesh, VarCompareOpts{VarCompareOpts::RELATIVE, tol, floor});
+    auto res = compare_meshes(&mesh, &mesh2, opts, true);
+    if (res == OMEGA_H_SAME || (allow_superset && res == OMEGA_H_MORE))
+    {
+      cout << "\nSUCCESS! Two meshes (.osh and .bp) are the same\n";
+      return 0;
+    }
+    cout << "\nFAIL! Two meshes (.osh and .bp) are NOT the same\n";
+     return 2;
+   }
+   catch (std::exception &e)
+   {
+     std::cout << "\nERROR: ADIOS2 exception: " << e.what() << "\n";
+#if ADIOS2_USE_MPI
+     MPI_Abort(lib.world()->get_impl(), -1);
+#endif
+   }
+    return 0;
+}
+
+// serial only at the moment
+void print_info(Library* lib, Omega_h::Mesh mesh)
+{
+  auto rank = lib->world()->rank();
+  
+  ostringstream oss;
+  // always print two places to the right of the decimal
+  // for floating point types (i.e., imbalance)
+  oss.precision(2);
+  oss << std::fixed;
+
+  if (!rank) 
+  {
+    oss << "\nEntity Type: " << Omega_h::topological_singular_name(mesh.family(), mesh.dim()) << "\n";
+
+    oss << "\nEntity Count and Imbalance: (Dim, #Global, #Local, Max/Ave Imbalance)\n";
+    for (int dim=0; dim <= mesh.dim(); dim++)
+      oss << "(" << dim << ", " 
+  	  << mesh.nglobal_ents(dim) << ", " 
+    	  <<mesh.nents(dim)<< ", " 
+	  << mesh.imbalance(dim) << ")\n";
+
+    oss << "\nTag by Dimension: (Name, Dim, Type, #Components, Min/Max Value)\n";
+    for (int dim=0; dim <= mesh.dim(); dim++)
+      for (int tag=0; tag < mesh.ntags(dim); tag++) {
+        auto tagbase = mesh.get_tag(dim, tag);
+        if (tagbase->type() == OMEGA_H_I8)
```

Comment:
You can use the `apply_to_omega_h_types` function which will avoid code duplication here. Function is defined here: https://github.com/SCOREC/omega_h/blob/00131eb0133796f39ff2ecd0962d9b3d14d5887b/src/Omega_h_defines.hpp#L86

example use case: https://github.com/SCOREC/omega_h/blob/00131eb0133796f39ff2ecd0962d9b3d14d5887b/src/Omega_h_rcFields.cpp#L335-L339